src/rust/aggregator/src/data/user_history.rs

```diff
@@ -0,0 +1,411 @@
+            )
+            .execute(&self.pool)
+            .await
+            .map_err(|e| DataAggregationError::ProcessingError(anyhow!(e)))?;
+        }
+        if !market_events.is_empty() {
+            for x in market_events {
+                sqlx::query!(
+                    r#"
+                        INSERT INTO aggregator.aggregated_events VALUES (
+                            $1, $2
+                        );
+                    "#,
+                    x.txn_version,
+                    x.event_idx,
+                )
+                .execute(&self.pool)
+                .await
+                .map_err(|e| DataAggregationError::ProcessingError(anyhow!(e)))?;
+            }
+        }
+        for x in &swap_events {
+            sqlx::query!(
+                r#"
+                    INSERT INTO aggregator.user_history_swap VALUES (
+                        $1, $2, $3, $4, $5, $6, $7, $8, $9
+                    );
+                "#,
+                x.market_id,
+                x.order_id,
+                x.direction,
+                x.limit_price,
+                x.signing_account,
+                x.min_base,
+                x.max_base,
+                x.min_quote,
+                x.max_quote,
+            )
+            .execute(&self.pool)
+            .await
+            .map_err(|e| DataAggregationError::ProcessingError(anyhow!(e)))?;
+            let market = sqlx::query!("SELECT * FROM market_registration_events WHERE market_id = $1", x.market_id)
+            .fetch_one(&self.pool)
+            .await
+            .map_err(|e| DataAggregationError::ProcessingError(anyhow!(e)))?;
+            sqlx::query!(
+                r#"
+                    INSERT INTO aggregator.user_history VALUES (
+                        $1, $2, $3, $4, $5, $6, $7, $8, $9
+                    );
+                "#,
+                x.market_id,
+                x.order_id,
+                x.time,
+                None as Option<DateTime<Utc>>,
+                x.integrator,
+                BigDecimal::zero(),
+                x.max_base.clone() / market.lot_size,
+                OrderStatus::Open as OrderStatus,
+                OrderType::Swap as OrderType,
+            )
+            .execute(&self.pool)
+            .await
+            .map_err(|e| DataAggregationError::ProcessingError(anyhow!(e)))?;
+        }
+        if !swap_events.is_empty() {
+            for x in swap_events {
+                sqlx::query!(
+                    r#"
+                        INSERT INTO aggregator.aggregated_events VALUES (
+                            $1, $2
+                        );
+                    "#,
+                    x.txn_version,
+                    x.event_idx,
+                )
+                .execute(&self.pool)
+                .await
+                .map_err(|e| DataAggregationError::ProcessingError(anyhow!(e)))?;
+            }
+        }
+        for x in &change_events {
```

Comment:
I suspect that with this you may get a size underflow if an order is created at 30, then changes size to 50, then fills 45, then changes size to 60, then fills 60. The total filled size is 105 but the size you'll have in the user history before you process fill events is 60.

I believe that you have to process/aggregate the events in the order that they happen. The txn version and event idx form a total ordering so that should help you.

Reply:
The underflow isn't a problem IMO. If the events are indexed in order, that means the underflow will only last inside the transaction, but will be gone once the transaction is done (DB transactions). If they are not, then ordering or not ordering will not make a difference because, but still, the underflow might last max 5 secs between aggregator scans. We can talk more about this on a call.